src/core_landice/mode_forward/mpas_li_calving.F

```diff
@@ -1276,170 +1249,411 @@ end subroutine eigencalving
+
+      ! 2. Calculate calving for dynamic cells
+      calvingSubtotal2 = 0.0_RKIND
+
+      ! a. Calculate calving on dynamic margin edges
+      do iCell = 1, nCellsSolve
+         if ( ( (li_mask_is_floating_ice(cellMask(iCell)) .and. li_mask_is_dynamic_ice(cellMask(iCell)) ) & ! floating dynamic
+                .or.  (groundedMarineMarginMask(iCell) == 1) ) &
+              .and. li_mask_is_margin(cellMask(iCell)) ) then
+            ! only consider dynamic margins adjacent to open ocean (need to exclude dynamic margins adjacent to land)
+            do iNeighbor = 1, nEdgesOnCell(iCell)
+               jCell = cellsOnCell(iNeighbor, iCell)
+               if ((.not. li_mask_is_ice(cellMask(jCell))) .and. (bedTopography(jCell) <= config_sea_level)) then
+                  iEdge = edgesOnCell(iNeighbor, iCell)
+                  edgeLengthScaling = scale_edge_length(angleEdge(iEdge), uReconstructX(1,iCell), uReconstructY(1,iCell))
+                  requiredCalvingVolumeDynEdge(iEdge) = calvingVelocity(iCell) * &
+                          edgeLengthScaling * dvEdge(iEdge) * thickness(iCell) * deltat
+               endif
+            enddo
+         endif
+      enddo
+
+      ! b. copy calving remaining in non-dynamic cells to dynamic edges
+      ! Assume height and velocity are uniform, but edge length is not.
+      do iCell = 1, nCellsSolve
+         if ( (.not. li_mask_is_dynamic_ice(cellMask(iCell))) .and. &
+                 uncalvedVolumeNonDynCell(iCell) > 0.0_RKIND) then
+            ! This is a non-dynamic location that still has calving to offer - a location where calving needs to be propagated
+
+            !call mpas_log_write("Passing calving from non-dynamic cell $i.  $r available to pass", &
+            !        realArgs=(/uncalvedVolumeNonDynCell(iCell)/), intArgs=(/iCell/))
+
+            volumeAvailableToPass = uncalvedVolumeNonDynCell(iCell)
+
+            ! Find total length of interface with dynamic cells at this cell
+            calvLengthCell = 0.0_RKIND
+            do iNeighbor = 1, nEdgesOnCell(iCell)
+               iEdge = edgesOnCell(iNeighbor, iCell)
+               if (li_mask_is_dynamic_ice(edgeMask(iEdge))) then
+                  edgeLengthScaling = scale_edge_length(angleEdge(iEdge), uReconstructX(1,iCell), uReconstructY(1,iCell))
+                  calvLengthEdge = edgeLengthScaling * dvEdge(iEdge)
+                  calvLengthCell = calvLengthCell + calvLengthEdge
+               endif
+            enddo
+
+            ! Now pass along the required calving volume relative to the interface length
+            do iNeighbor = 1, nEdgesOnCell(iCell)
+               iEdge = edgesOnCell(iNeighbor, iCell)
+               jCell = cellsOnCell(iNeighbor, iCell)
+               if (li_mask_is_dynamic_ice(edgeMask(iEdge))) then
+                  edgeLengthScaling = scale_edge_length(angleEdge(iEdge), uReconstructX(1,iCell), uReconstructY(1,iCell))
```

Comment:
This seems redundant. Isn't it exactly the same as the do-loop in lines 1052–1509?

Reply:
The first loop is to calculate calvLengthCell, which is needed before the second loop can be evaluated.